proposals/table-component.md

```diff
@@ -0,0 +1,221 @@
+# Table Component
+
+Component to render tabular data.
+
+## Features
+
+* Sorting
+
+## Currently Unsupported
+
+* Pagination
+* Filtering
+* Fixed row
+* Fixed column
+* Striped rows
+* Hoverable row styling
+* Multi-column sort
+* Row headers
+* Expandable row
+
+## Usage
+
+### Basic usage
+
+```javascript
+import React from 'react';
+import {Table} from 'baseui/table';
+
+const data = [
+  [1, 'John', 'Smith'],
+  [2, 'Lois', 'Lane'],
+];
+
+const columns = {
+  {
+    title: 'Employee ID',
+    dataIndex: 1,
```

Comment:
What is the `dataIndex` used for? Could `columns` be just an array?

Reply:
i considered relying on position in the array, but it's likely that the data may not be ordered in the way it should be rendered. similarly, a column may be a combination of multiple pieces of data from the provided row

Reply:
I think I understand it now - `dataIndex` serves as a selector for `data`. Then, this makes sense to me. 

We could also consider to make the `table` more "dumb" and let our users to slice and dice the data before they pass it in.  I feel that `dataIndex` and `render` construct doesn't save that much work unless I am missing something - the same thing can be achieved with something as 

```js
data={data.map(x => x.slice(1,4))}
```

or with re-ordering

```js
data={data.map(x => [x[3], x[1], x[2]])}
```


Also, it's one less API to learn/remember and would make our component simpler.

Reply:
cool, i like this pattern for getting rid of `dataIndex`. does this replace render? is something like below what you have in mind?

```jsx
<Table
  data={data.map(x => [
    (<Cell>{x[1]}</Cell>),
    (<CustomCell>{x[2]}</CustomCell>),
  ])}
/>
```

Reply:
That's a different thing and valid concern. `render` for column would give you option to swap the cell component for a specific column but what if I want to customize a specific row? Or one of the head cells? I can do overrides only for all head cells or all rows, right? I think we can generalize tables into these two patterns:

## Build your own table

```jsx
import {
  StyledRoot
  StyledHead
  StyledHeadCell
  StyledBody
  StyledRow
  StyledCell
} from 'baseui/table';

export default () =>
  <StyledRoot>
    <StyledHead>
      <StyledHeadCell>Column 1</StyledHeadCell>
      <StyledHeadCell>Column 2</StyledHeadCell>
    </StyledHead>
    <StyledBody>
      <StyledRow>
        <StyledCell>Foo</StyledCell>
        <StyledCell>Baz</StyledCell>
      </StyledRow>
    </StyledBody>
  </StyledRoot>;
```

This way, the user can customize/replace any node in the whole table. If there is no need for that, they can use `TableBuilder`.

## Use the table builder

```jsx
import { TableBuilder } from 'baseui/table';

export default () =>
  <TableBuilder 
    columns={['Column 1', 'Column 2']}
    data={[['Foo', 'Baz']]}
  />;
```

Internally, `TableBuilder` uses components from the previous example. It saves you a lot of imports and typing.

The idea is that `TableBuilder` should cover like 80% use-cases (the API can be a bit more permissive to achieve that). Also, you can still use our overrides to change styles for the whole class of components (for example, you could modify all `StyledHeadCells` if you wish).

In the rest 20%, we give you atomic components to build-your-own-thing.

The `render` prop for columns or this

```jsx
<TableBuilder
  data={data.map(x => [
    (<Cell>{x[1]}</Cell>),
    (<CustomCell>{x[2]}</CustomCell>),
  ])}
/>
```

seems like something in between those two worlds. It's hard to tell how much it's going to be used and makes our component/API pretty complex. I would make the first iteration as simple as possible - exporting atomic elements and minimalistic `TableBuilder` and add more features later.

**TLDR;** I am pretty much just describing what's already being proposed here. My only concern is if we want to immediately start mixing `TableBuilder` (or just `Table` in our proposal) with pass-your-own-components or should we just keep it more locked down, so you can pass only the data. There' s a lot of unsupported features we will probably want to eventually add and more complex the API is now, harder it will get.

Reply:
thanks for putting so much thought into this response, and I agree. let's try to get a simple table implementation out and consider additional features as the needs arise. will update the rfc